frontend/lib/uast/BuilderResult.cpp

```diff
@@ -190,90 +190,154 @@ ID BuilderResult::idToParentId(ID id) const {
+// M: <number of id-to-location entries, uint64_t>
+// <file path for locations, std::string>
+//   0..M-1:
+//     <id as key, ID>
+//     <first line, int>
+//     <first column, int>
+//     <last line, int>
+//     <last column, int>
+// <comment-id-to-location, std::vector<Location>>
+// <end sentinel string, std::string>
+void BuilderResult::serialize(Serializer& ser) const {
+  ser.write(DYNO_BUILDER_RESULT_START_STR);
+  ser.write(filePath_);
   const uint32_t numEntries = numTopLevelExpressions();
   ser.write(numEntries);
 
   for (auto ast : topLevelExpressions()) {
     ast->serialize(ser);
   }
+
+  ser.write(idToParentId_);
+
+  {
+    // TODO: For config-vars set on the command line, the 'file path' is set to
+    // something different. This leads to issues when comparing against a
+    // deserialized BuilderResult.
+    ser.write((uint64_t)idToLocation_.size());
+    ser.write(filePath_);
+    for (const auto& pair : idToLocation_) {
+      ser.write(pair.first);
+      ser.write(pair.second.firstLine());
+      ser.write(pair.second.firstColumn());
+      ser.write(pair.second.lastLine());
+      ser.write(pair.second.lastColumn());
+    }
+  }
+
+  ser.write(commentIdToLocation_);
+  ser.write(DYNO_BUILDER_RESULT_END_STR);
 }
 
-// TODO: handle Locations
-AstList BuilderResult::deserialize(Context* context, std::string& sfname) {
-  std::ifstream myFile;
-  myFile.open(sfname, std::ios::in | std::ios::binary);
+static void assignIDsFromTree(llvm::DenseMap<ID, const AstNode*>& idToAst,
+                              const AstNode* node) {
+  if (node->isComment()) return;
 
-  return deserialize(context, myFile);
+  idToAst[node->id()] = node;
+  for (auto child : node->children()) {
+    assignIDsFromTree(idToAst, child);
+  }
 }
 
-AstList BuilderResult::deserialize(Context* context, std::istream& is) {
-  AstList ret;
-  Deserializer des(context, is);
+BuilderResult BuilderResult::deserialize(Deserializer& des) {
+  BuilderResult ret;
+  AstList alist;
 
-  auto m = des.read<uint64_t>();
-  auto v = des.read<uint32_t>();
-  (void)m; // silence unused variable warnings
-  (void)v; // silence unused variable warnings
-  assert(m == magic);
-  assert(v == version);
+  CHPL_ASSERT(DYNO_BUILDER_RESULT_START_STR == des.read<std::string>());
+
+  auto path = des.read<UniqueString>(); // path
 
   const auto numEntries = des.read<uint32_t>();
 
+  // TODO: for improved performance, try recomputing the IDs rather than
+  // serializing and deserializing them. If we recompute these IDs then we
+  // also don't need to store the 'idToParent' map.
   for (uint32_t i = 0; i < numEntries; i++) {
-    ret.push_back(AstNode::deserialize(des));
+    alist.push_back(AstNode::deserialize(des));
+  }
+
+  auto idToParent = des.read<llvm::DenseMap<ID,ID>>();
+
+  // TODO: For performance and reduced file-size, try variable-byte encoding.
+  // TODO: For performance, could try not loading this data until it's needed.
```

Comment:
Does this mean you imagine we'll keep the file stream around with a deserializer over it until we want to read this?

Reply:
Yeah, just an idea.